feat: integration of frigate ephemeral scanning

sdmg15 opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Description

This PR integrates the experimental Frigate scanning server with ephemeral client keys.


### Notes to the reviewers

The specification and how to run a frigate server can be found at: https://github.com/sparrowwallet/frigate
### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [conventional commit guidelines](https://www.conventionalcommits.org/en/v1.0.0/)
* [x] I ran `just p` (fmt, clippy and test) before committing

sdmg15 committed:
feat(oracles): integrate frigate ephemeral scanning

nymius reviewed (commented):
I have tried the command without success. Could you provide me some examples to try?

If you could create an executable step by step example like the built in `doc/tabconf7` would be great.

Notice it builds a container with frigate on it. You only have to expose the ports to be able to communicate with it from outside of the container.

nymius commented on cli/v2/src/main.rs:
```diff
@@ -567,6 +577,95 @@ async fn main() -> anyhow::Result<()> {
+                        block.txdata = txs;
+                        tracing::debug!("Final block {:?}", block);
+                        wallet.apply_block_relevant(&block, secret.1, secret.0);
+                    }
```
Maybe we should improve the wallet interface to avoid this, but it works.
You're missing the chain update to make the wallet aware of the block this transaction belongs to. If not, the balance of transactions is going to be unconfirmed for ever.
```rust
let checkpoint = wallet.chain().tip().insert(BlockId { height, hash });
wallet.update_chain(checkpoint);
```

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,222 @@
+}
+
+const SUBSCRIBE_RPC_METHOD: &str = "blockchain.silentpayments.subscribe";
+const _UNSUBSCRIBE_RPC_METHOD: &str = "blockchain.silentpayments.unsubscribe";
```
We don't have long enough wait times in regtest to suffer from request taking minutes, but I would add a sane default to cut execution if no progress has been made in enough time.

nymius commented on cli/v2/src/main.rs:
```diff
@@ -567,6 +577,95 @@ async fn main() -> anyhow::Result<()> {
+                            .or_insert(HashMap::from([(h.tx_hash, h.tweak_key)]));
+                    });
+
+                    // Filter when the height is 0, because that would mean mempool transaction
```
I would add a comment mentioning this paragraph from `Frigate's` README, and how we are not following that approach here, and what this alternative is about.

sdmg15 replied:
I'm not sure I fully get the comment. But it's referenced here https://github.com/sparrowwallet/frigate/tree/master?tab=readme-ov-file#notifications

> height: The integer height of the block the transaction was confirmed in. For mempool transactions, 0 should be used.

nymius replied:
Sorry, I forgot to paste the paragraph:
> Clients should retrieve the transactions listed in the history with blockchain.transaction.get and subscribe to all owned outputs with blockchain.scripthash.subscribe. Electrum wallet functionality then proceeds as normal. In other words, the silent payments address subscription is a replacement for the monotonically increasing derivation path index in BIP32 wallets. The subscription seeks only to add to the client's knowledge of incoming silent payments transactions. The client is responsible for checking the transactions do actually send to addresses it has keys for, and using normal Electrum wallet synchronization techniques to monitor for changes to these addresses. The tweak key is provided to allow the client to avoid looking up the scriptPubKeys of spent outputs.

sdmg15 replied:
Yes we're already getting the transactions details with `blockchain.transaction.get` the only thing we're not doing here is `blockchain.scripthash.subscribe` because cmiiw we're just doing one time scanning so I didn't find it necessary to subscribe to the scripthash.

nymius replied:
I know, but I say it will be better to document the difference with the suggested approach.

nymius replied:
Also, this gave me an idea for a greater architectural change that may be worth exploring. To keep a queue of the blocks matched, and any other data given by the sp oracle (like txhash), and synchronize on demand by requesting this blocks to different sources. This would create a two stage synchronization process for light clients:
- Find matching blocks: frigate, blindbit + cbf, rpc
- Request matching blocks and finalize scanning: electrum, cbf, rpc, p2p

sdmg15 replied:
Yeah interesting. We could discuss further on that once this one is merged and I can give it a try it that's fine by you.

Btw will give you the full reports on what's preventing me from using the playbook and address the comments in the next days. Thanks.

nymius reviewed (commented):
I fixed the issue with `bdk-tx`. You should rebase.

Successfully executed scanning on regtest:

```bash
$ REGTEST_ADDRESS=$(regtest-bdk unused_address | jq -r '.address' | tr -d '\n')
$ just mine 1 $REGTEST_ADDRESS
[
  "5e724d83a5e8999874ec679ed5823962f56c82bf4810be937b248dc5a08b15cb"
]
$ just mine 101 2>&1 >/dev/null
$ regtest-bdk sync
{}
$ regtest-bdk balance
{
  "satoshi": {
    "confirmed": 5000000000,
    "immature": 0,
    "trusted_pending": 0,
    "untrusted_pending": 0
  }
}
$ SP_CODE=$(regtest-sp code | jq -r '.silent_payment_code' | tr -d '\n')
$ RAW_TX=$(regtest-bdk create_sp_tx --to-sp $SP_CODE:10000 --fee_rate 5 | jq -r '.raw_tx' | tr -d '\n')
$ TXID=$(regtest-bdk broadcast --tx $RAW_TX | jq -r '.txid' | tr -d '\n')
$ just mine 1
[
  "2c28f3fa2cdd742a3e5e564f770802c558324506ea6db3ab6c0cea85ab83c0c3"
]
$ regtest-bdk sync
{}
$ regtest-bdk balance
{
  "satoshi": {
    "confirmed": 4999989228,
    "immature": 0,
    "trusted_pending": 0,
    "untrusted_pending": 0
  }
}
$ regtest-sp scan-frigate --url 127.0.0.1:57001 --start 0
2026-02-12T19:10:12.162809Z  WARN bdk_sp_oracles::frigate: Read bytes from stream 59
2026-02-12T19:10:12.183998Z  WARN bdk_sp_oracles::frigate: Read bytes from stream 156
2026-02-12T19:10:12.184062Z  INFO bdk_sp_oracles::frigate: Subscribed to silent payment address: String("sprt1qqvf42gn2770nwf8jvalczzv80ad8p9zseahnleklzzklqxkg5s80wqa7xxx58l2wvumjlmrzy9eed4vsd647jcmp6jg5tr08unyhex7u5chnkk64")
2026-02-12T19:10:12.184092Z  INFO sp_cli2: Starting frigate scanning loop...
2026-02-12T19:10:12.361669Z  WARN bdk_sp_oracles::frigate: Read bytes from stream 461
2026-02-12T19:10:12.414640Z  WARN bdk_sp_oracles::frigate: Read bytes from stream 200
2026-02-12T19:10:12.474964Z  WARN bdk_sp_oracles::frigate: Read bytes from stream 448
2026-02-12T19:10:12.475701Z  INFO sp_cli2: Progress 1
2026-02-12T19:10:12.475740Z  WARN sp_cli2: Scanning completed
$ echo $SP_CODE
sprt1qqvf42gn2770nwf8jvalczzv80ad8p9zseahnleklzzklqxkg5s80wqa7xxx58l2wvumjlmrzy9eed4vsd647jcmp6jg5tr08unyhex7u5chnkk64
$ regtest-sp balance
{
  "confirmed": {
    "immature": 0,
    "spendable": 0,
    "total": 0
  }
}
{
  "unconfirmed": {
    "total": 10000,
    "trusted": 0,
    "untrusted": 10000
  }
}
```

Notice the balance appears as unconfirmed because the block to which the transaction is anchored doesn't belong to the wallet local chain. I left a comment in the code mentioning how to fix that.

I think that `Read bytes from stream {bytes}` belongs to the `DEBUG` scope, rather than `WARN`. I would also print a one line log of the balance, after each execution of the command.

Could you give me more information about the issues that are preventing you from testing with the playbooks? Could you setup nix?

Good work!

sdmg15 commented:
Added a new playbook https://github.com/bitcoindevkit/bdk-sp/pull/48/commits/59a1128b72cf7192cd84c9ef0f58edab3993c9ff. It's important to note that for this work we need to update Frigate to 1.3.2

nymius commented on cli/v2/src/main.rs:
```diff
@@ -567,6 +578,146 @@ async fn main() -> anyhow::Result<()> {
                 );
             }
         }
+        Commands::ScanFrigate { rpc_args, start } => {
```
For `start` you could implement something similar than `ScanCbf` logic:

```rust
let sync_point = if let (Some(height), Some(hash)) = (height, hash) {
    HeaderCheckpoint::new(height, hash)
} else if wallet.birthday.height <= wallet.chain().tip().height() {
    let height = wallet.chain().tip().height();
    let hash = wallet.chain().tip().hash();
    HeaderCheckpoint::new(height, hash)
} else {
    let checkpoint = wallet
        .chain()
        .get(wallet.birthday.height)
        .expect("should be something");
    let height = checkpoint.height();
    let hash = checkpoint.hash();
    HeaderCheckpoint::new(height, hash)
};
```

nymius commented on cli/v2/src/main.rs:
```diff
@@ -567,6 +578,146 @@ async fn main() -> anyhow::Result<()> {
             }
         }
+        Commands::ScanFrigate { rpc_args, start } => {
+            // The implementation done here differ from what mentionned in the section https://github.com/sparrowwallet/frigate/tree/master?tab=readme-ov-file#blockchainsilentpaymentssubscribe
```
```suggestion
// The implementation done here differs from what is mentioned in the section
// https://github.com/sparrowwallet/frigate/tree/master?tab=readme-ov-file#blockchainsilentpaymentssubscribe
// This implementation is doing a one time scanning only. So instead of calling
// `blockchain.scripthash.subscribe` on each script from the wallet, we just subscribe
// and read the scanning result from the stream. On each result received we update the
// wallet state and once scanning progress reaches 1.0 (100%) we stop.
```

nymius commented on cli/v2/src/main.rs:
```diff
@@ -567,6 +578,146 @@ async fn main() -> anyhow::Result<()> {
+            let mut client = FrigateClient::connect(&rpc_args.url)
+                .await
+                .unwrap()
+                .with_timeout(tokio::time::Duration::from_secs(600));
```
10 minutes is a lot of time. I would reduce this to 60 seconds.

nymius commented on cli/v2/src/main.rs:
```diff
@@ -567,6 +578,146 @@ async fn main() -> anyhow::Result<()> {
+                                raw_blk.push_str("00");
+
+                                // Push dummy coinbase
+                                let dummy_coinbase = "01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff1b03951a0604f15ccf5609013803062b9b5a0100072f425443432f20000000000000000000";
```
I would make a const of this hex value.

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,296 @@
+const UNSUBSCRIBE_RPC_METHOD: &str = "blockchain.silentpayments.unsubscribe";
+const GET_RPC_METHOD: &str = "blockchain.transaction.get";
+const VERSION_RPC_METHOD: &str = "server.version";
+const _SUBSCRIBE_OWNED_OUTPUTS: &str = "blockchain.scripthash.subscribe";
```
I would remove this if unused.

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,296 @@
+            self.client.write_all(req_bytes).await?;
+            self.client.write_all(b"\n").await?;
+            self.client.flush().await?;
+            self.read_from_stream(4096).await
```
Why 4096? Because it is the default TCP packet size? I would make it a const.

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,296 @@
+    pub async fn get_block_header(&mut self, height: u32) -> Result<String, FrigateError> {
+        let params = vec![height];
+        let req = RequestPayload {
+            method: "blockchain.block.header".to_string(),
```
You haven't defined a const for this method.

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,296 @@
+        let req = RequestPayload {
+            method: "blockchain.block.header".to_string(),
+            params: serde_json::json!(params),
+            id: serde_json::Value::from(1000),
```
What's the purpose for this ID? Shouldn't be unique for each request? Should it be the same for related request? Does it make sense to have it as a parameter?

sdmg15 replied:
Yes it's a required parameter by frigate server, else we won't get any reply. Yes it could be the same for related requests. Given that we're doing a one time scan having it static is still fine. I initially just chose number on top of my head, but I'll update to use increasing sequence.

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,296 @@
+        let req = RequestPayload {
+            method: GET_RPC_METHOD.to_string(),
+            params: serde_json::json!(params),
+            id: serde_json::Value::from(51),
```
Same question about ID than before.

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,296 @@
+        let req = RequestPayload {
+            method: VERSION_RPC_METHOD.to_string(),
+            params: serde_json::json!(params),
+            id: serde_json::Value::from(71),
```
Also about ID.

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,296 @@
+        let req = RequestPayload {
+            method: SUBSCRIBE_RPC_METHOD.to_string(),
+            params: serde_json::json!(params),
+            id: serde_json::Value::from(91),
```
Also about ID.

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,296 @@
+        let req = RequestPayload {
+            method: UNSUBSCRIBE_RPC_METHOD.to_string(),
+            params: serde_json::json!(params),
+            id: serde_json::Value::from(189),
```
About ID.

nymius commented on Cargo.lock:
```diff
@@ -366,6 +367,8 @@ dependencies = [
  "rand 0.8.5",
  "rand_core 0.6.4",
+ "rand 0.8.5",
+ "rand_core 0.6.4",
```
Why are you repeating dependencies here? Remove this from the commit, please.

nymius reviewed (changes requested):
I've checked the `frigate_playbook.sh` execution and it works great. I'm glad you were able to replicate it.

I recommend you to rebase and check each commit individually and:
- execute `git show --check` and fix any issue arising.
- execute `just p` and implement fixes and apply formatting.
- rewrite commits to use [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/). Look at commit history to understand the conventions and scopes used.
- squash the commits into a single one.

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,299 @@
+            method: SUBSCRIBE_RPC_METHOD.to_string(),
+            params: serde_json::json!(params),
+            id: serde_json::Value::from(2),
+            //:Value::from(91),
```
nit: this shouldn't be here

nymius commented on cli/v2/src/main.rs:
```diff
@@ -161,6 +163,16 @@ pub enum Commands {
+        #[clap(long)]
+        hash: Option<BlockHash>,
+    },
+
```
Remove the new line here.

nymius commented on cli/v2/src/main.rs:
```diff
@@ -161,6 +163,16 @@ pub enum Commands {
         #[clap(long)]
         hash: Option<BlockHash>,
     },
+
```
Remove the new line here. Commands do not have padding between them.

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,298 @@
+const VERSION_RPC_METHOD: &str = "server.version";
+const BLOCK_HEADER_RPC_METHOD: &str = "blockchain.block.header";
+const STREAM_READ_BYTES: usize = 4096;
+pub const DUMMY_COINBASE: &str = "01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff1b03951a0604f15ccf5609013803062b9b5a0100072f425443432f20000000000000000000";
```
Move these below imports

```rust
use bip157::tokio::io::{AsyncReadExt, AsyncWriteExt};
use bip157::tokio::net::TcpStream;
use bip157::tokio::time::{timeout, Duration};
use bitcoin::secp256k1::{PublicKey, SecretKey};
use bitcoin::Txid;
use serde::{Deserialize, Serialize};
use serde_json::Value;

const SUBSCRIBE_RPC_METHOD: &str = "blockchain.silentpayments.subscribe";
const UNSUBSCRIBE_RPC_METHOD: &str = "blockchain.silentpayments.unsubscribe";
const GET_RPC_METHOD: &str = "blockchain.transaction.get";
const VERSION_RPC_METHOD: &str = "server.version";
const BLOCK_HEADER_RPC_METHOD: &str = "blockchain.block.header";
const STREAM_READ_BYTES: usize = 4096;
pub const DUMMY_COINBASE: &str = "01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff1b03951a0604f15ccf5609013803062b9b5a0100072f425443432f20000000000000000000";
```

nymius commented on oracles/src/frigate/mod.rs:
```diff
@@ -0,0 +1,298 @@
+
+    pub async fn get_block_header(&mut self, height: u32) -> Result<String, FrigateError> {
+        let params = vec![height];
+        let req = RequestPayload {
```
I have researched a little bit more about JSON-RPC, and it seems id must be unique per each request and it is used to correlate responses with the original request. This is specifically desinged for cases like this, where you are working with multiple asynchronous calls and the order is not guaranteed. Look here: https://www.jsonrpc.org/specification#parameter_structures.
Therefore, this is incomplete:
- Ids should be unique
- You must check the id matches the response id.
- What happens if `response["error"]` is not empty?

nymius commented:
I've pushed fixes and new changes to work on this from the nix environment, please rebase.

sdmg15 committed:
feat: use electrum_streaming_client for handling requests

sdmg15 commented:
@nymius I addressed the comments in a separate commit https://github.com/bitcoindevkit/bdk-sp/pull/48/commits/7c1cd0569dbc2c47f8c294f0904ad53305d18d68. Once approved I'll squash.

nymius reviewed (changes requested):
Hi @sdmg15, I think we should simplify this and avoid implementing JSON-RPC
code as much as possible. Your approach is working, locally, but the protocol
has some nuances that should be carefully considered.

I suggest you to have a look at [`electrum_streaming_client`](https://github.com/bitcoindevkit/electrum_streaming_client), it has:

- [notification](https://www.jsonrpc.org/specification#notification) management:
    - Treates notifications as a different thing than responses:
      https://github.com/bitcoindevkit/electrum_streaming_client/blob/master/src/state.rs#L143,
      https://github.com/bitcoindevkit/electrum_streaming_client/blob/master/src/client.rs#L168-L194
- JSON encoding/decoding is done
  following [JSON: RFC 4627](https://www.ietf.org/rfc/rfc4627.txt) as specified by JSON-RPC protocol, in its [overview](https://www.jsonrpc.org/specification#overview):
    - Expects JSON encoded bytes in buffers and doesn't treat TCP block chunks
      indifferently:
      https://github.com/bitcoindevkit/electrum_streaming_client/blob/master/src/io.rs#L86
- Requests and responses are tracked and matched by their id, as
  [expected](https://www.jsonrpc.org/specification#response_object):
  https://github.com/bitcoindevkit/electrum_streaming_client/blob/master/src/state.rs#L97-L128
  In your code, you're expecting the latest response to automatically match the
  id of the last request, which is going to work because you're not sending
  multiple request at the same time, but then, what's the point of implementing
  an asynchronous client anyway?

I recommend you to register the Frigate endpoints in a separate fork of
`electrum_streaming_client` (as a gated feature) and consider carefully how to
match the code conventions of the host repository. Then you can use that code
as a git dependency here.

sdmg15 commented:
Alright thanks for checking @nymius, I'll do that and let you know the outcome.

nymius commented on oracles/Cargo.toml:
```diff
@@ -14,6 +14,7 @@ rayon = "1.11.0"
 reqwest = { version = "0.12.23", features = ["json", "rustls-tls", "http2", "charset"], default-features = false }
 serde = { version = "1.0.219", features = ["serde_derive"] }
 serde_json = { version = "1.0.142", features = ["raw_value"]}
+electrum_streaming_client = { path = "/home/sdmg15/workspace/electrum_streaming_client"}
```
You should use a git path here, not a file path. Is not going to work on a different host.

sdmg15 replied:
Oopss sorry this was while doing local testing, I forgot to update in this file

nymius commented:
> By stable you mean once the opened PR there is merged?

Yes, or requested changed don't require breaking changes.
